### PR TITLE
All-In-One grinder results are now defined by type, not by huge lists

### LINKED ISSUE #32369: All-In-One Grinders only grinding one instead of all
During round 78272 I noticed as chef that my all-in-one grinder was only grinding one reagent at a time instead of all of them at once. With a whole ton of wheat needed to mass-produce my cakes this was obviously super frustrating and time-consuming; I don't know much about PR's or coding, but I can't seem to find any notes saying this was intentionally added.

### LINKED ISSUE #32532: Grindables dont grind
With the new grindables anonmare pr, I wanted to make some meth. I tried grinding down some glowsticks and lighttubes, and got nothing. It didnt even fucking grind.

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: Grinding runed metal and brass now produces iron/blood and iron/teslium, respectively.
balance: As part of some code-side improvements, the amount of reagents you get from grinding some objects might be slightly different.
fix: Some grinding recipes that didn't work, like dead mice and glowsticks, now do.
fix: All-In-One grinders now correctly grind up everything, instead of one thing at a time.
/:cl:

Fixes #32532. (I think!)
Fixes #32369.

I had some time free, and noticed how awful the reagent grinder code was - it used huge static lists containing types and their associated reagents from grinding.

This is now split into two new vars on `/obj/item` - `var/list/grind_results` and `var/list/juice_results`, as well as two new helper procs, `on_grind()` and `on_juice()` to allow those to change based on conditions like plant potency. Such checks and the like have been moved to that. If any of these procs return -1, the operation is canceled.

I also fixed some of the recipes that didn't work. The reagent IDs for them didn't exist, leading me to believe that they weren't tested. I corrected that! (I've tested every single recipe in this PR, with the exception of a few juicing-related ones.)

### CODE REVIEW

**Comment by optimumtact** [comment]
this should really be a grindable component datum that has the two vars on it, but i cant find it in my heart to care quite that much

**Comment by Xhuis** [comment]
Adding an Initialize() to some ~40-50 objects just to make them grindable also seems a tad silly (unless Initialize's point is to do that, of course.)

**Comment by optimumtact** [comment]
It's called anyway, you're just doing extra setup work

**Comment by Xhuis** [comment]
You know, I promise I'm capable of higher thought.

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -676,6 +680,15 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
 
 /obj/item/proc/on_mob_death(mob/living/L, gibbed)
 
+/obj/item/proc/grind_requirements(obj/machinery/reagentgrinder/R) //Used to check for extra requirements for grinding an object
+	return TRUE
+
+ //Called BEFORE the object is ground up - use this to change grind results based on conditions
+ //Use "return -1" to prevent the grinding from occurring
+/obj/item/proc/on_grind(obj/machinery/reagentgrinder/R) //Note that it isn't necessary to include the argument if you don't mess with the reagent grinder at all
```
Comment (Cyberboss) [inline code/game/objects/items.dm]: Remove that last comment, we don't condone messing with base proc args without good reason

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
I do agree with @optimumtact on the component thing

File: code/game/objects/items.dm (hunk 1 above)
Comment (Xhuis) [reply code/game/objects/items.dm]: It occurs to me, actually, that I currently don't use the argument in any instance of it, so I believe I'll snip it out unless something has need of it.

File: code/game/objects/items/stacks/sheets/mineral.dm [hunk 2]
```diff
@@ -127,6 +127,7 @@ GLOBAL_LIST_INIT(diamond_recipes, list ( \
 	sheettype = "uranium"
 	materials = list(MAT_URANIUM=MINERAL_MATERIAL_AMOUNT)
 	novariants = TRUE
+	grind_results = list("uranium" = 2)
```
Comment (XDTM) [inline code/game/objects/items/stacks/sheets/mineral.dm]: are these actually 2u of chems per sheet or am i missing something?

Comment (Xhuis) [reply code/game/objects/items/stacks/sheets/mineral.dm]: There are, same for all the mineral sheet types.

Comment (XDTM) [reply code/game/objects/items/stacks/sheets/mineral.dm]: Shouldn't it be 20 per sheet to match current gameplay? I'm sure if you grind a chair's worth of metal you can get more than a few drops of iron

Comment (Xhuis) [reply code/game/objects/items/stacks/sheets/mineral.dm]: Hm, I thought it was just a flat 20. If it's 20 currently, I can do that, sure.

**Comment by Doppeltongue** [comment]
For growns, how does this interact with using different plants to produce special reagents? For instance, I usually splice chlorine, radium, and phosphorus into wheat to use wheat's good stats for producing mutagen, but now it seems like even if I take out all the vitamin and nutriment, I will end up with some amount of flour in the grinder.

File: code/game/objects/items.dm [hunk 3]
```diff
@@ -676,6 +680,15 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
 
 /obj/item/proc/on_mob_death(mob/living/L, gibbed)
 
+/obj/item/proc/grind_requirements(obj/machinery/reagentgrinder/R) //Used to check for extra requirements for grinding an object
+	return TRUE
+
+ //Called BEFORE the object is ground up - use this to change grind results based on conditions
+ //Use "return -1" to prevent the grinding from occurring
+/obj/item/proc/on_grind()
```
Comment (coiax) [inline code/game/objects/items.dm]: These functions should, by default, return `grind_results` and `juice_results` or a -1 if the list is null or empty. 

That way, subtypes can override these functions for more complicated stuff, like returning the reagents list but modified

File: code/game/objects/items/stacks/sheets/mineral.dm (hunk 2 above)
Comment (Doppeltongue) [reply code/game/objects/items/stacks/sheets/mineral.dm]: Right now, grinding produces 20u of chems per sheet. For instance, grinding one sheet of plasma in xenobiology produces 20 u of plasma, which is enough for 4 reactions. If this really results in 2 units of plasma, then xenobiology, chemistry, and virology are going to have issues.

**Comment by Xhuis** [comment]
@Doppeltongue Grinding it up will do the same thing as it currently does - make the special grind reagent, in addition to the spliced ones.

**Comment by Doppeltongue** [comment]
@Xhuis Right now, I can remove all possibility of flour appearing and contaminating the grind results by removing vitamin and nutriment, so no, that's not how it currently works.

**Comment by Xhuis** [comment]
Ohhhh, I see what you mean. I believe the formula to do that was implemented before such a feature, and that it's unintentional as a result. I may reimplement it.

File: code/game/objects/items/apc_frame.dm [hunk 4]
```diff
@@ -111,7 +111,7 @@
 	return TRUE
 
 
-/obj/item/electronics
+/obj/item/electronics //WHY IS THIS DEFINED HERE?
```
Comment (lzimann) [inline code/game/objects/items/apc_frame.dm]: remove this useless comment

**Comment by JJRcop** [comment]
If you componentize this, add an on_grind callback var that is invoked by the grinder when it grinds something.

File: code/game/objects/items/stacks/sheets/mineral.dm (hunk 2 above)
Comment (Ratvar) [reply code/game/objects/items/stacks/sheets/mineral.dm]: You need 1 unit of plasma for reaction to start, no? 20u of plasma is 20 reactions, if you are careful

Comment (Supermichael777) [reply code/game/objects/items/stacks/sheets/mineral.dm]: yeah but your going to want more that 2 reactions befor you need to bow and scrape to the guy running R&D/Robo

**Comment by Xhuis** [comment]
I'm getting mixed signals on if this should be componentized; should it?

**Comment by optimumtact** [comment]
It should, but given that it'd be a two var component realistically I can't demand you do it, so it's up to you

**Comment by Xhuis** [comment]
Hmm, I'll mull it over.

File: code/game/objects/items.dm (hunk 3 above)
Comment (Xhuis) [reply code/game/objects/items.dm]: `on_grind()` exists in order *to* change the reagents list somehow before grinding the object. What you described seems to be the already-existing function of the proc.

**Comment by Xhuis** [comment]
I tweaked the amount of reagents you get from plants to be like the old formula, where it's based on how much nutriment is in the plant. This means you can splice out the flour from wheat, etc. like before this change.

**Comment by Cyberboss** [comment]
I'm working on componentizing the implementation which I'm gonna PR to your branch

**Comment by Xhuis** [comment]
Should I fix the conflicts, or will you do that in your commit?

File: code/game/objects/items/stacks/sheets/sheet_types.dm [hunk 5]
```diff
@@ -347,6 +351,7 @@ GLOBAL_LIST_INIT(brass_recipes, list ( \
 	throw_range = 3
 	turf_type = /turf/open/floor/clockwork
 	novariants = FALSE
+	grind_results = list("iron" = 0.5, "teslium" = 1.5)
```
Comment (swindly) [inline code/game/objects/items/stacks/sheets/sheet_types.dm]: But brass is made of copper (and zinc). There's no iron in it.

**Comment by Cyberboss** [comment]
I'll take care of it @Xhuis

File: code/game/objects/items/stacks/sheets/sheet_types.dm (hunk 5 above)
Comment (Xhuis) [reply code/game/objects/items/stacks/sheets/sheet_types.dm]: "Brass sheets" are actually just replicant alloy fused into the sturdy form that looks like brass but is strong enough to build with (normal brass probably wouldn't make good solid walls.)

**Comment by Cyberboss** [comment]
Sorry for the delay, I'll make it a priority tonight

**Comment by Xhuis** [comment]
No worries!

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Juicing amounts used to be based on a proc called get_juice_amount. With your changes as is everything juiced will produce 0 reagents, could you fix that bit please?

**Comment by Cyberboss** [comment]
Go ahead and fix the merge conflicts too while you're at it and I'll rebase

File: code/game/objects/items/stacks/bscrystal.dm [hunk 6]
```diff
@@ -62,6 +62,7 @@
 	novariants = TRUE
 	grind_results = list("bluespace" = 2)
 	var/crystal_type = /obj/item/stack/ore/bluespace_crystal/refined
+	var/crystal_type = /obj/item/ore/bluespace_crystal/refined
```
Comment (Cyberboss) [inline code/game/objects/items/stacks/bscrystal.dm]: :?

Comment (Xhuis) [reply code/game/objects/items/stacks/bscrystal.dm]: That's odd...

**Comment by optimumtact** [comment]
Is this still WIP?

**Comment by optimumtact** [comment]
I'm not going to hold this for you @Cyberboss

**Comment by Cyberboss** [comment]
Yeah I'm dragging my heels I'll fix it tonight or tomorrow

**Comment by Xhuis** [comment]
image

**Comment by JJRcop** [comment]
@Cyberboss

**Pull request merged**